src/physics/site_types/electron.jl

```diff
@@ -0,0 +1,107 @@
+export ElectronSite,
+       electronSites
+
+function electronSites(N::Int; kwargs...)
+  return [Index(4,"Site,Electron,n=$n") for n=1:N]
+end
+
+const ElectronSite = makeTagType("Electron")
+
+function state(::ElectronSite,
+               st::AbstractString)
+  if st == "0" || st == "Emp"
+    return 1
+  elseif st == "Up" || st == "↑"
+    return 2
+  elseif st == "Dn" || st == "↓"
+    return 3
+  elseif st == "UpDn" || st == "↑↓"
+    return 4
+  end
+  throw(ArgumentError("State string \"$st\" not recognized for Electron site"))
+  return 0
+end
+
+function op(::ElectronSite,
+            s::Index,
+            opname::AbstractString)::ITensor
+  sP = prime(s)
+  Emp   = s(1)
+  EmpP  = sP(1)
+  Up    = s(2)
+  UpP   = sP(2)
+  Dn    = s(3)
+  DnP   = sP(3)
+  UpDn  = s(4)
+  UpDnP = sP(4)
+
+  Op = ITensor(dag(s), s')
+
+  if opname == "Nup"
```

Comment:
This may be a bit of an aside from this pull request, but can we use a similar trick to use dispatch to choose the operator instead of a long list of if statements?

For example, we could do:

```julia
function op(::ElectronSite, s::Index, opname::AbstractString)
  Op = ITensor(dag(s), s')
  op!(Op, s, ElectronSite(), Val(Symbol(opname)))
  return Op
end

function op!(Op::ITensor, s::Index, ::ElectronSite, ::Val{:Nup})
  Op[s(2),s'(2)] = 1.
  Op[s(4),s'(4)] = 1.
  return Op
end
```

The advantage of this is avoiding a very long set of `if ... elseif ...` statements. Also, this allows people to add their own operators to an existing site type by overloading `op!` (in case it is something complicated we may not want to include, but is useful for their own use case).

If we don't like `Val` in the signature, it may be nicer to use a type alias like `const OpType{T} = Val{T}` so then the signature is `OpType{:Nup}`. We could also use Tag/SmallString instead of a symbol, but right now SmallString doesn't allow unicode (and of course has length limits).

Reply:
Yes I was thinking about something like this too, or at least using this trick more broadly. I wish this kind of thing was more natively built into the language hmm. 

I think we could always add this later without having to do any breaking changes, though. So let's definitely keep it in mind. There are some details to think about, like whether to use a SmallString or a Symbol (good thing about symbol is it can be long, but it forces a somewhat advanced concept on users).

Probably a good solution is to create a helper function that returns Val(Symbol(opname)) so users can just call that.